### ISSUE TITLE
Dynamic link sometimes redirects to appstore

### ISSUE THREAD

**Issue by bobbyzamora** [firebase/firebase-ios-sdk#7822]
### Step 1: Describe your environment

  * Xcode version: 12.4
  * Firebase SDK version: 7.9.0
  * Installation method: CocoaPods
  * Firebase Component: Dynamic Links

### Step 2: Describe the problem
Hi!
We're using a lot of **firebase dynamic links** in our iOS application, and almost all time it works perfectly. 
But, sometimes opening application from link stops working: link opening redirects users to Appstore page.
It doesn't depend on:
- device or iOS version
- way of link opening (manually created link from firebase admin panel, from mail app, pasting directly to safari, opening from notes etc)
- app usage time (it can start wrong redirecting after week of daily app use, for example) 
- firebase version (we are fighting with this bug for about 6 month, and every app release we have a fresh sdk version).
- app launch type: debug from xcode, adhoc with testflight, or release from Appstore

And most important - **it doesn't depend on the link**. Like if you hook this bug while opening some link, everyone else using the same link will not have any problems.

So, it seems to us that the problem is inside **device or firebase sdk**.

The bright side is that this bug goes away by itself, after some amount of time, or rebooting/reinstalling.

How we are dealing with this problem:
prompting customers to:
 - past link directly to safari (doesn't help most time)
 - reboot device (helps most time)
 - reinstall app (helps most time)
 - use "trick" to past link to notes, long press to context menu, select "open in app" (helps most time)
 
If it matters, we have the same functionality in android application, and there wasn't any problems like this.

#### Steps to reproduce:

Unfortunately we can't provide any info about how to see this bug, because we can't reproduce it on our side. It just randomly stops working properly.

#### What our configs are look like:
https://ourappdomain/apple-app-site-association:
`{"applinks":{"apps":[],"details":[{"appID":"ourTeam.ourApp","paths":["NOT /_/*","/*"]}]}}`

OurApp.entitlements:
```
<key>com.apple.developer.associated-domains</key>
<array>
   <string>applinks:ourapp.page.link</string>
</array>
```

Info.plist:
```
<key>LSApplicationQueriesSchemes</key>
<array>
    <string>googlegmail</string>
    <string>ourapp.page.link</string>
</array>
<key>FirebaseDynamicLinksCustomDomains</key>
    <array>
      <string>https://ourapp.page.link</string>
    </array>
```

AppDelegate:
```
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]? = nil) -> Bool {
        FirebaseApp.configure()
...

```
```
    func application(
        _ application: UIApplication,
        continue userActivity: NSUserActivity,
        restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void) -> Bool {
        let handled = DynamicLinks.dynamicLinks().handleUniversalLink(userActivity.webpageURL!) { (dynamiclink, error) in
            if let url = dynamiclink?.url?.absoluteString {
                DeepLinkHandler.handleLink(url)
            }
        }

        return handled
    }
```

```
  func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
        return true
  }
```

Thanks in advance for any help!

**Comment by rizafran** [comment]
Hi @bobbyzamora. I see that the link is redirecting to App Store even if the app is installed, however, you’re unable to reproduce it on your end. For us to verify the root cause of the issue or to isolate it, can you provide the following details:
- Affected device (OS version/device model)
- App wherein the Dynamic Link has been opened
- Dynamic Link used
    - Is it only happening to 1 Dynamic Link?
- Pattern/s you’ve noticed on the issue (ex. only happening to iOS 13)
- Confirmation if there are no warnings when appending ?d=1 to your affected link ([Debugging Dynamic Links](https://firebase.google.com/docs/dynamic-links/debug))

Thanks.

**Comment by bobbyzamora** [comment]
Hi @rizafran! Thanks for the reply!
Here is some info:

**1. Affected device (OS version/device model):**
iPhone XS, iOS 14.4
iPhone 12 pro max, iOS 14.4
iPhone 6s, iOS 14.3
iPhone 6s, iOS 14.4
iPhone 12 Pro, iOS 14.4.2 & iOS 14.4.1
iPhone 8, iOS 14.4.2

(this is not the full list, just some examples from our customers reports)

**2. App wherein the Dynamic Link has been opened**
Didn't quite get what kind of info do you need here:)
We have an iOS app and Appstore page. Usually dynamic link redirects to app, but sometimes to Appstore.

**3. Dynamic Link used. Is it only happening to 1 Dynamic Link?**
Nope, it doesn't depend on the link. If you encounter this bug, all links (generated by api, or created manually through the firebase admin panel - doesn't matter) will stop opening correctly. And same links will work properly on any other device.

**4. Pattern/s you’ve noticed on the issue (ex. only happening to iOS 13)**
We have plenty a lot of customers complaints about links not opening in the app, but unfortunately didn't spot any patterns. So it really looks like a "random temporary bug".

**5. Confirmation if there are no warnings when appending ?d=1 to your affected link (Debugging Dynamic Links).**
We haven't done this kind of debug yet. Thanks for this link. We gonna try it when find this bug again and i write what kind of logs it has.

Thanks again rizafran, hope this info will help somehow.

**Comment by rizafran** [comment]
Thanks for sharing the info, @bobbyzamora. The issue seems to be happening intermittently, and I wasn't able to reproduce it on my end. In case you have reproduced the issue or if you find something that would help with the investigation, including the [debugging of your Dynamic Link](https://firebase.google.com/docs/dynamic-links/debug) (adding ?d=1), feel free to share it here. Thanks.

**Comment by bobbyzamora** [comment]
Hi @rizafran! I'm returning with some info. I've just catch this error, and here is what output i have.

Launched app with Xcode12.4 on iPhone 11, iOS 14.4.2. Got the link that redirects to AppStore, then added:
`DynamicLinks.performDiagnostics(completion: nil)`
and 
`suffix?d=1` to the link

--------------
Xcode console output:

```
[Firebase/Crashlytics] Version 7.9.0
7.9.0 - [Firebase/Analytics][I-ACS023007] Analytics v.7.9.0 started
7.9.0 - [Firebase/Analytics][I-ACS023008] To enable debug logging set the following application argument: -FIRAnalyticsDebugEnabled (see http://goo.gl/RfcP7r)

WF: _userSettingsForUser mobile: {
    filterBlacklist =     (
    );
    filterWhitelist =     (
    );
    restrictWeb = 1;
    useContentFilter = 0;
    useContentFilterOverrides = 0;
    whitelistEnabled = 0;
}
WF: _WebFilterIsActive returning: NO
7.9.0 - [Firebase/Messaging][I-FCM001000] FIRMessaging Remote Notifications proxy enabled, will swizzle remote notification receiver handlers. If you'd prefer to manually integrate Firebase Messaging, add "FirebaseAppDelegateProxyEnabled" to your Info.plist, and set it to NO. Follow the instructions at:
https://firebase.google.com/docs/cloud-messaging/ios/client#method_swizzling_in_firebase_messaging
to ensure proper integration.


---- Firebase Dynamic Links diagnostic output start ----
Firebase Dynamic Links framework version 7.9.0
System information: OS iOS, OS version 14.4.2, model iPhone
Current date 2021-04-22 15:48:22 +0000
Device locale en-RU (raw en_RU), timezone Europe/Moscow
	Specified custom URL scheme is thestoicapp and Info.plist contains such scheme in CFBundleURLTypes key.
	AppID Prefix: D4P55P6FV6, Team ID: D4P55P6FV6, AppId Prefix equal to Team ID: YES
performDiagnostic completed successfully! No errors found.
---- Firebase Dynamic Links diagnostic output end ----

<Warning>: Please set a value for FacebookAutoLogAppEventsEnabled. Set the flag to TRUE if you want to collect app install, app launch and in-app purchase events automatically. To request user consent before collecting data, set the flag value to FALSE, then change to TRUE once user consent is received. Learn more: https://developers.facebook.com/docs/app-events/getting-started-app-events-ios#disable-auto-events.

<Warning>: You haven't set a value for FacebookAdvertiserIDCollectionEnabled. Set the flag to TRUE if you want to collect Advertiser ID for better advertising and analytics results.
FBSDKLog: Cannot load gate keepers before configuring.

7.9.0 - [Firebase/Analytics][I-ACS800023] No pending snapshot to activate. SDK name: app_measurement
AppsFlyer SDK version 6.2.4 started build (71);
Keychain items migration status: 0 
7.9.0 - [Firebase/Analytics][I-ACS023012] Analytics collection enabled
7.9.0 - [Firebase/Analytics][I-ACS023220] Analytics screen reporting is enabled. Call +[FIRAnalytics logEventWithName:FIREventScreenView parameters:] to log a screen view event. To disable automatic screen reporting, set the flag FirebaseAutomaticScreenReportingEnabled to NO (boolean) in the Info.plist


 [assertion] Error acquiring assertion: <Error Domain=RBSAssertionErrorDomain Code=2 "Specified target process does not exist" UserInfo={NSLocalizedFailureReason=Specified target process does not exist}>
[ProcessSuspension] 0x116507730 - ProcessAssertion: Failed to acquire RBS Background assertion 'WebProcess Background Assertion' for process with PID 19789, error: Error Domain=RBSAssertionErrorDomain Code=2 "Specified target process does not exist" UserInfo={NSLocalizedFailureReason=Specified target process does not exist}

[Client] {"msg":"#NullIsland Either the latitude or longitude was exactly 0! That's highly unlikely", "latIsZero":0, "lonIsZero":0}
[Client] {"msg":"#NullIsland Received a latitude or longitude from getLocationForBundleID that was exactly zero", "latIsZero":0, "lonIsZero":0, "location":{"floor":2147483647,"lifespan":-1,"rawLat":0,"integrity":0,"referenceFrame":"Unknown","lon":0,"speed":-1,"type":"Unknown","altitude":0,"rawCourse":-1,"confidence":0,"suitability":{"type":"decode failure","raw value":451406408,"expected type":"CLClientLocationSuitability"},"timestamp":-1,"rawReferenceFrame":"Unknown","lat":4.9406564584124654e-324,"verticalAccuracy":-1,"rawLon":0,"horizontalAccuracy":-1,"speedAccuracy":-1,"courseAccuracy":-1,"course":-1}}
```
--------------

Then i've tried opening the link: "https://stoicapp.page.link/hkqdPBE3sHBxjins8" that starts redirecting me to the AppStore.
Then i've added debug suffix to the link, and here is what i got:

![dl_debug](https://user-images.githubusercontent.com/2696734/115746877-63884500-a39d-11eb-8b83-051953401588.png)

--------------

After this, i've made copy-paste link to the notes app. Press the link from the notes, and it redirected successfully to the app.

-------------

To sum up, nothing new with links behaviour from my previous messages (sometimes all links starts redirecting to the AppStore, after opening from notes app - everything becomes ok). I see some errors from console, but dynamic links logs looks fine.

Thanks in advance!

**Comment by rizafran** [comment]
Thanks for sharing your findings, @bobbyzamora. I've tried debugging your dynamic link by adding a `?d=1` in the end, and I've noticed that there are 3 warnings. Could you double-check if the `UniversalLinks` is enabled and if you're using the correct bundle ID?

**Comment by bobbyzamora** [comment]
@rizafran Thanks for the reply! 
_Universal link_ is enabled, and _bundle id_ is correct. Otherwise it would not open later from notes app.
Note that after opening from "notes" the link can be opened from safari too.
Thanks!

**Comment by rizafran** [comment]
Hi @bobbyzamora, can you file a ticket to [Firebase support](https://firebase.google.com/support)? This way, you can share your project info to the support team so that they might help you to troubleshoot your dynamic link further. Thanks.
